language/diem-tools/prefetched-transaction-replay/src/lib.rs

```diff
@@ -0,0 +1,410 @@
+        ensure_dir_does_not_exist(&config.build_dir)?;
+        ensure_dir_does_not_exist(&config.storage_dir)?;
+
+        let debugger = DiemDebugger::json_rpc_with_config(
+            &config.rpc_url,
+            config.build_dir.clone(),
+            config.storage_dir.clone(),
+        )?;
+
+        let rpc_client = BlockingClient::new(&config.rpc_url);
+
+        // NOTE: we could probably avoid creating a second validator
+        // by making the one in `DiemDebugger` accessible publicly
+        let validator = Box::new(JsonRpcDebuggerInterface::new(&config.rpc_url)?);
+        Ok(PrefetchTransactionReplayer {
+            config,
+            validator,
+            debugger,
+            rpc_client,
+        })
+    }
+
+    fn get_keys_read_for_tx(
+        &self,
+        live_state: &impl MoveResolver,
+        tx: &Transaction,
+        rw_analysis: &ReadWriteSetAnalysis,
+    ) -> Result<Vec<ResourceKey>> {
+        let signed_tx = match tx {
+            Transaction::UserTransaction(u) => u,
+            t => {
+                warn!("cannot process tx {:?}, ignoring", t);
+                return Ok(vec![]);
+            }
+        };
+
+        // NOTE: we might need to do something about the transactions we do
+        // not currently handle
+        match signed_tx.payload() {
+            TransactionPayload::ScriptFunction(_) => {
+                rw_analysis.get_keys_read(signed_tx, live_state)
+            }
+            payload => {
+                warn!("cannot process payload {:?}, ignoring", payload);
+                Ok(vec![])
+            }
+        }
+    }
+
+    fn get_unique_addresses_from_keys(&self, key_reads: &[ResourceKey]) -> Vec<AccountAddress> {
+        key_reads
+            .iter()
+            .map(|k| k.address)
+            .collect::<HashSet<_>>()
+            .into_iter()
+            .collect()
+    }
+
+    fn extract_account_state(&self, state_with_proof: &AccountStateWithProofView) -> AccountState {
+        // NOTE: if this has a chance to fail, it might be better to change
+        // the function to return Result<AccountState>
+        let blob = state_with_proof.blob.as_ref().unwrap();
+        let parsed_blob = bcs::from_bytes::<AccountStateBlob>(blob).unwrap();
+        AccountState::try_from(&parsed_blob).unwrap()
+    }
+
+    fn fetch_accounts_storage(
+        &self,
+        addresses: &[AccountAddress],
+    ) -> Result<Vec<AccountAddressWithState>> {
+        let requests = addresses
+            .iter()
+            .map(|a| MethodRequest::get_account_state_with_proof(*a, None, None))
+            .collect();
+
+        let results = self
+            .rpc_client
+            .batch(requests)?
+            .into_iter()
+            .collect::<Result<Vec<_>, _>>()?;
+
+        let account_states: Vec<AccountAddressWithState> = results
+            .into_iter()
+            .zip(addresses.iter())
```

Comment:
I'm not familiar enough with the guarantees of the batch request, but are you guaranteed to get them back in the same order as the request batch?

Reply:
Good point! I doubled-check this, the sdk takes care of ordering the responses in the same order as the request batch: https://github.com/diem/diem/blob/main/sdk/client/src/lib.rs#L121